service/shadowsocks.go

```diff
@@ -44,7 +44,7 @@ type ServiceMetrics interface {
 
 type Service interface {
 	HandleStream(ctx context.Context, conn transport.StreamConn)
-	HandlePacket(conn net.PacketConn)
+	HandlePacket(conn net.Conn, pkt []byte)
```

Comment:
I don't see a benefit in moving the read loop out of the handler.

Make the handler take a UDP Association instead (net.Conn). This simplifies things. Like, it knows all packets in that association are from the same client and you can leverage the NAT from Caddy.

You also remove a layer of buffering. The shadowsocks code may be able to read the packed directly into its buffer. You change is perhaps doubling the buffer memory needed.